### PR TITLE
feat(FN-3664): add risk data to created customers

### PR DESCRIPTION
**Pull request by natdeanlewissoftwire**
## Introduction :pencil2:
In discussion with Risk teams, they wanted Risk data to be added to newly created customers in Salesforce.
The APIM side will default Risk rating at B+ and Loss given default at 50%, and the DTFS side will pass Probability of Default through to APIM to be set on the new record.

## Resolution :heavy_check_mark:
Defaults the first two risk values, and sets the Probability of Default value to the value DTFS passes through.

## Miscellaneous :heavy_plus_sign:
A small doc typo fix I found.

### CODE REVIEW

File: src/helpers/date-formatter.helper.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Returns the current date in the correct format for ingestion by the Salesforce sObject API.
```
Comment (abhi-markan) [inline src/helpers/date-formatter.helper.ts:2]:
```suggestion
 * Returns the current date in the correct format for ingestion by the Salesforce Object API.
```

File: src/helpers/date-formatter.helper.test.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  beforeEach(() => {
+    jest.spyOn(global, 'Date').mockImplementation(() => mockDate);
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  it('should return the current date, formatted correctly', () => {
```
Comment (abhi-markan) [inline src/helpers/date-formatter.helper.test.ts:14]: Can we add more test cases:

1. Past date 1970-01-01.
2. Future date 9999-99-99.
3. Leap year date 2020-02-29.

File: src/helpers/date-formatter.helper.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+ *
+ * @returns {string} the current date in yyyy-mm-dd format
+ */
+
+export const salesforceFormattedCurrentDate = () => {
+  const today = new Date();
+  const dd = String(today.getDate()).padStart(2, '0');
+  const mm = String(today.getMonth() + 1).padStart(2, '0');
+  const yyyy = String(today.getFullYear());
+  return `${yyyy}-${mm}-${dd}`;
```
Comment (abhi-markan) [inline src/helpers/date-formatter.helper.ts:12]:
```suggestion
  
  return `${yyyy}-${mm}-${dd}`;
```

File: src/modules/customers/customers.service.ts [hunk 4]
```diff
@@ -205,6 +206,10 @@ export class CustomersService {
       Party_URN__c: partyUrn,
       D_B_Number__c: dunsNumber,
       Company_Registration_Number__c: DTFSCustomerDto.companyRegistrationNumber,
+      CCM_Credit_Risk_Rating__c: 'B+',
```
Comment (abhi-markan) [inline src/modules/customers/customers.service.ts:209]: Can this be a constant please.

File: src/modules/customers/customers.service.ts [hunk 5]
```diff
@@ -205,6 +206,10 @@ export class CustomersService {
...
+      CCM_Credit_Risk_Rating_Date__c: salesforceFormattedCurrentDate(),
+      CCM_Loss_Given_Default__c: 50,
```
Comment (abhi-markan) [inline src/modules/customers/customers.service.ts:211]: Can this be a constant please.

File: src/modules/customers/dto/create-customer.dto.ts [hunk 6]
```diff
@@ -25,4 +25,28 @@ export class CreateCustomerDto {
...
   @MaxLength(10)
   Company_Registration_Number__c: string;
+
+  @ApiProperty({ description: 'Credit Risk Rating' })
+  @IsString()
+  @IsNotEmpty()
+  CCM_Credit_Risk_Rating__c: string;
+
+  @ApiProperty({ description: 'Credit Risk Rating Date' })
+  @IsString()
```
Comment (abhi-markan) [inline src/modules/customers/dto/create-customer.dto.ts:35]: Will this be in `YYYY-MM-DD` format, if yes can we set the length decorator to `10`?

File: src/modules/customers/dto/create-customer.dto.ts [hunk 7]
```diff
@@ -25,4 +25,28 @@ export class CreateCustomerDto {
...
+  @IsNotEmpty()
+  CCM_Credit_Risk_Rating_Date__c: string;
+
+  @ApiProperty({ description: 'Loss Given Default' })
+  @IsNumber()
+  @IsNotEmpty()
+  @Min(0)
+  @Max(100)
```
Comment (abhi-markan) [inline src/modules/customers/dto/create-customer.dto.ts:43]: Is the maximum `100`, want to make sure this can also be a float number or a string with percentage sign?

File: src/modules/customers/dto/create-customer.dto.ts [hunk 8]
```diff
@@ -25,4 +25,28 @@ export class CreateCustomerDto {
...
+  CCM_Loss_Given_Default__c: number;
+
+  @ApiProperty({ description: 'Probability of Default' })
+  @IsNumber()
+  @IsNotEmpty()
+  @Min(0)
+  @Max(100)
```
Comment (abhi-markan) [inline src/modules/customers/dto/create-customer.dto.ts:50]: * Is the maximum `100` or `100.00`?

File: src/modules/customers/dto/dtfs-customer.dto.ts [hunk 9]
```diff
@@ -13,4 +13,9 @@ export class DTFSCustomerDto {
   @IsString()
   @IsNotEmpty()
   companyName: string;
+
+  @ApiProperty({ description: 'Probability of Default' })
+  @IsNumber()
+  @IsNotEmpty()
```
Comment (abhi-markan) [inline src/modules/customers/dto/dtfs-customer.dto.ts:19]: Will this accept `14.1` as a default PoD from DTFS?

File: src/modules/salesforce/salesforce.service.test.ts [hunk 10]
```diff
@@ -105,6 +109,10 @@ describe('SalesforceService', () => {
         Party_URN__c: null,
         D_B_Number__c: null,
         Company_Registration_Number__c: '12341234',
+        CCM_Credit_Risk_Rating_Date__c: '2007-03-27',
+        CCM_Credit_Risk_Rating__c: 'B+',
+        CCM_Loss_Given_Default__c: 50,
+        CCM_Probability_of_Default__c: null,
```
Comment (abhi-markan) [inline src/modules/salesforce/salesforce.service.test.ts:115]: Can we add some assertions with edge values.
PoD: 100
Risk Rating: AAA+
LGD: 100

File: test/docs/__snapshots__/get-docs-yaml.api-test.ts.snap [hunk 11]
```diff
@@ -808,9 +808,13 @@ components:
         companyName:
           type: string
           description: Company Name
+        probabilityOfDefault:
+          type: number
+          description: Probability of Default
```
Comment (abhi-markan) [inline test/docs/__snapshots__/get-docs-yaml.api-test.ts.snap:813]: Is this missing LGD and customer rating?

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: src/helpers/date-formatter.helper.ts (hunk 1 above)
Comment (natdeanlewissoftwire) [reply src/helpers/date-formatter.helper.ts:2]: Funnily enough this is actually not a typo - I think the s stands for salesforce?
https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_methods_system_sobject.htm

File: src/modules/customers/dto/create-customer.dto.ts (hunk 6 above)
Comment (natdeanlewissoftwire) [reply src/modules/customers/dto/create-customer.dto.ts:35]: Sure, good shout

File: src/modules/customers/dto/dtfs-customer.dto.ts (hunk 9 above)
Comment (natdeanlewissoftwire) [reply src/modules/customers/dto/dtfs-customer.dto.ts:19]: Will add a test for this

File: test/docs/__snapshots__/get-docs-yaml.api-test.ts.snap (hunk 11 above)
Comment (natdeanlewissoftwire) [reply test/docs/__snapshots__/get-docs-yaml.api-test.ts.snap:813]: Ah thanks I'll regenerate

File: src/modules/customers/dto/create-customer.dto.ts (hunk 7 above)
Comment (natdeanlewissoftwire) [reply src/modules/customers/dto/create-customer.dto.ts:43]: I'm enforcing this to be a float/number not a string which I think is reasonable. That's how it's stored in DTFS anyway so seems to be the most consistent choice. Floats pass these conditons as to plain integers.

File: src/modules/customers/dto/create-customer.dto.ts (hunk 8 above)
Comment (natdeanlewissoftwire) [reply src/modules/customers/dto/create-customer.dto.ts:50]: see above, 100.00 passes this, 100 is fine too. Will add some tests for this behaviour.
